### ISSUE TITLE
Support incremental map/reduce

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#1118]
We had this on our radar for a while, but didn't have an issue to track it. Since some people have been asking for an official issue to track, I'm adding this to GitHub.

I'm going to write up a specific proposal a bit later. This is in backlog as it's obviously a medium-term priority feature.

**Comment by bjornua** [comment]
:+1:

**Comment by natew** [comment]
Any word on the progress of this?

**Comment by danielmewes** [comment]
@natew This is definitely coming, but we don't have a specific release for it yet.

**Comment by coffeemug** [comment]
I think we'll be able to get this into 2.3 (although as @danielmewes mentioned, there is no ETA yet). Tentatively, I'd expect this 2-4 months from now.

**Comment by v3ss0n** [comment]
So current solution for count is to keep track of counts in separate field? 
for example , a new message arrived to `room` , it need to be inserted to count field of `channel` table. RIght? where changes is listening.

**Comment by danielmewes** [comment]
@v3ss0n You mean as a work-around until we implement incremental map/reduce?
Yeah, that sounds about right. We're also still planning to implement constant-time `count` (#152), but that too is a few releases away.

**Comment by v3ss0n** [comment]
Thanks i am doing that way ,. but that will cause 2 writes .

**Comment by thelinuxlich** [comment]
For inspiration: https://docs.influxdata.com/influxdb/v0.10/query_language/continuous_queries/

**Comment by danielmewes** [comment]
Interesting. I need to take a look at them.
On first sight, InfluxDB's continuous queries seem more specialized, but would be a very interesting use case.

**Comment by meenie** [comment]
Any movement on this?

**Comment by danielmewes** [comment]
@meenie Not yet. This is probably going to be the next thing after https://github.com/rethinkdb/rethinkdb/issues/3997 .

**Comment by danielmewes** [comment]
This has actually become really easy, due to the addition of the `fold` term. We can now formulate this entirely as a set of rewrites.

Given a changefeed of the form:
```js
stream.reduce(f).changes({includeInitial: <II>, includeStates: <IS>})
```

Assume that for the given `f`, we know the following properties:

* `<f_BASE>` the initial accumulator for `f`
* `<f_APPLY>` a function from the accumulator and an element in the input table to a new accumulator
* `<f_UNAPPLY>` the inverse of `<f_APPLY>` in the accumulator
* `<f_EMIT>` generates a result value of the reduction from the current accumulator

Now the query can be rewritten into:
```js
stream.changes({includeInitial: true, includeStates: true}).fold(
  {f_acc: <f_BASE>, is_initialized: false}),
  function(acc, el) {
    var f_acc = acc('f_acc');
    var new_f_acc = r.branch(el.hasFields("old_val"), <f_UNAPPLY>(f_acc, el('old_val')), acc).do(function(un_f_acc) {
        return r.branch(el.hasFields("new_val"), <f_APPLY>(un_f_acc, el('new_val')), un_f_acc);
      });
    var new_is_initialized = acc('is_initialized').or(el.hasFields('state').and(el('state').eq('ready')));
    return {f_acc: new_f_acc, is_initialized: new_is_initialized};
  },
  {emit: function(old_acc, el, new_acc) {
    var old_val = <f_EMIT>(old_acc);
    var new_val = <f_EMIT>(new_acc);
    // We handle the 'ready' state separately below
    var emit_state = r.expr(<IS>).and(el.hasFields('state')).and(el('state').ne('ready'));
    var emit_update = old_acc('is_initialized').and(old_val.ne(new_val));
    var emit_initial = r.expr(<II>).and(old_acc('is_initialized').not().and(new_acc('is_initialized')));
    return r.branch(
      emit_state, [el],
      emit_update, [{'old_val': old_val, 'new_val': new_val}],
      emit_initial, r.branch(<IS>, [{'new_val': new_val}, {state: "ready"}], [{'new_val': new_val}]),
      []
    );
  }})
```

For example for `count()`:

* `<f_BASE> = 0`
* `<f_APPLY> = function(acc, el) { return acc.add(1); }`
* `<f_UNAPPLY> = function(acc, el) { return acc.sub(1); }`
* `<f_EMIT> = function(acc) { return acc; }`

Or for `avg()`:

* `<f_BASE> = {c: 0, sum: 0}`
* `<f_APPLY> = function(acc, el) { return {c: acc('c').add(1), sum: acc('sum').add(el) }; }`
* `<f_UNAPPLY> = function(acc, el) { return {c: acc('c').sub(1), sum: acc('sum').sub(el) }; }`
* `<f_EMIT> = function(acc) { return acc('sum').div(acc('c')); }` (plus some sort of handling for empty input sets that we need to come up with)

The main disadvantage of this implementation is that it doesn't distribute the reduction anymore and causes additional network traffic between the shards and the parsing node. This can be improved through a bit of special code.

It would be amazing I think if we could get the slow version into 2.4, at least for the built-in terms (such as `count`, `sum` and `avg`). We can mark it as a "preview" feature, since it doesn't have the full scalability yet that you'd expect from RethinkDB, and then ship the optimized version with 2.5.

**Comment by meenie** [comment]
@danielmewes The initial reduction would be distributed, so the speed is still there.  Subsequent reductions would be very very cheap, so it wouldn't matter if they were distributed or not.  This is pretty awesome :).

**Comment by danielmewes** [comment]
Right, that would be the optimized version with an initial reduction. The simple one that works entirely as a rewrite to changes.fold would not distribute the initial reduction.

**Comment by meenie** [comment]
Oh, sorry, I misunderstood.  Awesome either way :).

**Comment by v3ss0n** [comment]
I Love the second one!
so at 2.4 , we can do counts on changefeed? that will be awesome!

**Comment by danielmewes** [comment]
> so at 2.4 , we can do counts on changefeed? that will be awesome!

Yes, that's the idea :)
The initial count will run a bit slower in a changefeed for 2.4 compared to running just the `count` without `changes`. But we can optimize this in the future to get almost the same performance.

**Comment by meenie** [comment]
If I could do a changefeed on an aggregate that [calculates NPS](https://www.getfeedback.com/examples/net-promoter-score), that would be amazing.  I have data that looks something like this:
```javascript
const npsData = [
  {
    "component_id": 1,
    "number": 10
  },
  {
    "component_id": 1,
    "number": 10
  },
  {
    "component_id": 2,
    "number": 8
  },
  {
    "component_id": 1,
    "number": 9
  },
  {
    "component_id": 2,
    "number": 2
  },
  ...
];
```

And my query looks something like this:
```javascript
r.expr(data)
  .group('component_id', 'number').count()
  .ungroup()
  .map((row) => {
    const number = row('group').nth(1);
    const ret = r.expr({
      component_id: row('group').nth(0),
      distribution: [{number: number, total: row('reduction')}],
      total_answers: row('reduction'),
      detractors: 0,
      passives: 0,
      promoters: 0
    });

    return r.branch(
      number.eq(9).or(number.eq(10)),
      ret.merge({promoters: ret('promoters').add(row('reduction'))}),
      number.eq(7).or(number.eq(8)),
      ret.merge({passives: ret('passives').add(row('reduction'))}),
      ret.merge({detractors: ret('detractors').add(row('reduction'))})
    );
  })
  .group('component_id')
  .reduce((left, right) => ({
    component_id: left('component_id'),
    total_answers: left('total_answers').add(right('total_answers')),
    detractors: left('detractors').add(right('detractors')),
    passives: left('passives').add(right('passives')),
    promoters: left('promoters').add(right('promoters')),
    distribution: left('distribution').add(right('distribution')),
  }))
  .do((datum) => {
    const passivesPercentage = datum('passives').div(datum('total_answers')).mul(100);
    const promotersPercentage = datum('promoters').div(datum('total_answers')).mul(100);
    const detractorsPercentage = datum('detractors').div(datum('total_answers')).mul(100);
    return {
      distribution: datum('distribution'),
      passives_percentage: passivesPercentage,
      promoters_percentage: promotersPercentage,
      detractors_percentage: detractorsPercentage,
      score: promotersPercentage.sub(detractorsPercentage)
    };
  })
  .ungroup()
  .map(row => ({
    component_id: row('group'),
    distribution: row('reduction')('distribution'),
    passives_percentage: row('reduction')('passives_percentage'),
    promoters_percentage: row('reduction')('promoters_percentage'),
    detractors_percentage: row('reduction')('detractors_percentage'),
    score: row('reduction')('score')
  }));
```

Could the above be achieved with this new api?

**Comment by v3ss0n** [comment]
@danielmewes  Excellent !

**Comment by danielmewes** [comment]
@meenie It depends on whether or not you can express it as a `reduce` operation and whether there's an efficient "reverse" function that updates the query result when a document gets removed from the input set.

**Comment by meenie** [comment]
@danielmewes So I wouldn't be able to use `group()`? And I'd have to do those counts manually using reduce? And ya, I believe you could reverse the above because you keep track of the distribution.

**Comment by meenie** [comment]
@danielmewes: Ya, that makes total sense.  For now, we need every bit of efficiency we can get, so I'll be experimenting with rewriting out queries to use changefeed's, but won't utilise this in production until it's on parity with speed.

**Comment by deontologician** [comment]
> r.table('test').filter(...).reduce{|a,b| ...}.changes(reverse: lambda {|acc, o| ...})
r.table('test').filter(...).reduce(reverse: lambda {|acc, o| ...}){|a,b| ...}

Since the reverse function is only needed because it's a changefeed, it seems like the first one makes sense. But then again it's not clear which function it's reversing if they're separated. I guess it feels kind of wrong to me to require something extra when you're doing `.changes` vs. a normal query.

**Comment by danielmewes** [comment]
>  I guess it feels kind of wrong to me to require something extra when you're doing `.changes` vs. a normal query.

The way I think of it is that this is like needing to have the `{index: ...}` optarg for `orderBy` if you want to have a changefeed on it. I can see that this is slightly different because the `reverse` option will not have any effect unless you open a changefeed, but I don't feel like that's a big issue.

I don't like the first one because it seems limiting and different from what we do anywhere else. 

What if in the future we allow changefeeds on queries that contain multiple `reduce` operations (for example within a subquery)? Specifying the function in `changes` would not work for that.

Or what if you have a query that looks like this: `tbl.reduce(...).do(...).changes()`? Ignoring the fact that we currently don't support `do` in changefeeds (which we totally should), it becomes much less obvious what the `reverse` argument to `changes` actually applies to and how it works. Does it get applied to the value after or before the `do`?

**Comment by deontologician** [comment]
Yeah, it seems like the best way is to provide the reverse function to the reduce term. I'm assuming if you don't tack on .changes the reverse optarg will just be a no-op (vs. erroring)?

**Comment by danielmewes** [comment]
> I'm assuming if you don't tack on .changes the reverse optarg will just be a no-op (vs. erroring)?

Yeah that's what I thought. That way you can run the same query with and without `.changes`.

**Comment by mlucy** [comment]
There's a lot of discussion above.  Here's my understanding the
current proposal:
* You can write any of these:
  - stream.avg(...).changes()
  - stream.sum(...).changes()
  - stream.count(...).changes()
  - stream.reduce(..., reverse: FUNC).changes()
* In particular, it doesn't need to be on a selection;
  `.map.reduce.changes` etc. are legal.  (We should probably support
  `concat_map.reduce.changes` even though we don't yet support
  `.concat_map.changes`, since it's easy.)

A few other things:
* Should we support `.coerce_to(...).changes()`?  The most common
  would probably be `.coerce_to('array').changes()`, where we'd
  re-send the whole array every time it changes.  There's an argument
  that coercing to an array is a terminal, so it might be more
  consistent to support it.
* Should we support `.group.reduce.changes`?  There's no real
  technical limitation, it would be almost as easy as not supporting
  it.  If so, should we also take this opportunity to support
  `.group.changes`?
* How should we handle reductions over nothing?
  (E.g. `r.table('test').avg('foo').changes()` when `test` changes
  from empty to non-empty -- what's `old_val`?)  Currently `sum` and
  `count` return 0 on empty streams, while `avg` and `reduce` produce
  an error.
  - We should probably use `0` as the "nothing" value for the
    terminals that return it on an empty stream.
  - One option would be to just use `nil` as the "nothing" value for
    all other terminals.
  - Another option would be to error by default, but to let people
    write e.g.  `.avg('foo').default(whatever).changes()` to specify
    it explicitly.

Also, on the subject of implementation, it probably wouldn't actually be
that hard to do it the efficient way where we do chunks of reductions
on the shards and only ship the aggregates over.  It would only speed
up the initial computation, but it would probably speed it up a *lot*.
(The reason I don't think it would be particularly hard is that we're
already only tracking timestamps on a per-read-transaction basis, so
we wouldn't lose any fidelity if we attached a terminal to the reads
we ship over and got back a pre-aggregated value alongside the
stamps.)

**Comment by v3ss0n** [comment]
`.coerce_to(...).changes()` looks like a convinience function , looks good but should be optional .
What i found exciting is `group.changes` and `group.reduce.changes` .

**Comment by danielmewes** [comment]
@mlucy Thanks for the summary of the current proposal. That matches what I had in mind for 2.4.

I'd like to add `coerce_to(...).changes()` from your suggestions to this as it seems trivial to do.

--

The three extensions that you're suggesting

1. `.coerce_to(...).changes()`
2. `.group(...).reduce(...).changes()`
3. more efficient implementation for the initial result

all sound really cool to me.

As far as I can tell, 1 would be easy to implement even as a pure `fold`-based rewrite at least to `coerceTo('array')`. We can just keep the current array in the accumulator. I'm not sure if there are any other types that we allow coercing to from a stream? `'string'` maybe? Most likely those would also be easy to support, and we would still maintain the array but then just call a final `.coerceTo(...)` on the accumulator array before emitting a value. In any case doing this will be `O(n)` in the number of results, but that's expected since the output per change is already of that size.

My impression is that 2 (`group.reduce.changes`) is a bit more involved in terms of having to figure out how to represent added and removed groups in the output stream.

Since we have limited remaining development resources for 2.4 considering the other things we are working on, my suggestion would be that we agree on a minimal proposal, and keep extensions 2 and 3 out of the proposal for now. If we end up having extra time, we can still implement the more efficient algorithm (3) or discuss grouped changefeeds separately.

--

> How should we handle reductions over nothing?

Great question.
My opinion is that we should emit them as the value `null` for `avg` and `reduce` (and `0` for `sum` and `count`). Our current changefeeds already use `null` to indicate the absence of a value. I think this would fit pretty nicely.
Reporting them as errors sounds nice on paper, but I think in practice it will be a much bigger pain for our users to handle.

**Comment by danielmewes** [comment]
Also I would like to add that I'm *extremely* excited about this feature! It's going to be so amazing :-)

**Comment by v3ss0n** [comment]
Can we have 2 and 3 in 2.5? :D

**Comment by danielmewes** [comment]
@v3ss0n I think so :)

**Comment by mlucy** [comment]
@danielmewes -- leaving 2 and 3 for later sounds good to me.  I don't think 2's representation would be a particularly involved discussion, though -- I was imagining we'd just emit the entire grouped data every time it changed (so `{old_val: {grp: red, grp2: red2}, new_val: {grp: red}}`).  If we wanted to support plain old `.group.changes` that would require thinking a little about the format, though.

On `coerce_to`, I think `coerce_to('array')` and `coerce_to('object')` are the only ones that can take a stream.

**Comment by danielmewes** [comment]
Marking settled as:

* You can write any of these:
  - stream.avg(...).changes()
  - stream.sum(...).changes()
  - stream.count(...).changes()
  - stream.reduce(..., reverse: FUNC).changes()
  - stream.coerceTo('array').changes()
  - stream.coerceTo('object').changes()
* In particular, it doesn't need to be on a selection;
  `.map.reduce.changes` etc. are legal.  (We should probably support
  `concat_map.reduce.changes` even though we don't yet support
  `.concat_map.changes`, since it's easy.)

For 2.4 we will implement the slower variant that performs the initial reduction on the parsing node rather than distributing it.

**Comment by marshall007** [comment]
This is gonna be great! Note that supporting `.coerce_to('array').changes()` solves the use-case I had for #3719, so I would say we probably don't need anything from that proposal anymore.

**Comment by nighelles** [comment]
This is in review 3714, except for coerce_to("array")

**Comment by dvergeylen** [comment]
As this have been transferred to milestone 2.4-polish (for obvious reasons), I just wanted to emphase what @danielmewes has written at [this comment](https://github.com/rethinkdb/rethinkdb/issues/1118#issuecomment-222234293) as a work-around for the time being. I suggest the following to support the other common aggregation operations:

* `sum()`: Very similar to `count()` but we increment by the value of `el` instead of 1.
	* `<f_BASE> = 0`
	* `<f_APPLY> = function(acc, el) { return acc.add(el); }`
	* `<f_UNAPPLY> = function(acc, el) { return acc.sub(el); }`
	* `<f_EMIT> = function(acc) { return acc; }`
* `min()`:
	* `<f_BASE> = Number.POSITIVE_INFINITY`
	* `<f_APPLY> = function(acc, el) { return (el < acc ? el : acc); }`
	* `<f_UNAPPLY> = function(acc, el) { return acc; }`
	* `<f_EMIT> = function(acc) { return acc; }`
* `max()`:
	* `<f_BASE> = Number.NEGATIVE_INFINITY`
	* `<f_APPLY> = function(acc, el) { return (el > acc ? el : acc); }`
	* `<f_UNAPPLY> = function(acc, el) { return acc; }`
	* `<f_EMIT> = function(acc) { return acc; }`
* `avg()`: Similar to what Daniel did, but I would handle empty sets with a one liner like this:
	* `<f_EMIT> = function(acc) { return (acc('c').neq(0) ? acc('sum').div(acc('c')) : 0); }`

Any thoughts on this would be appreciated. :smiley:
